### ISSUE TITLE
0001792:
add links to notification mails

### ISSUE THREAD

**Issue by Gloirin** [tine20/Tine-2.0-Open-Source-Groupware-and-CRM#980]
**Reported by pschuele on 15 Oct 2009 12:26**

add links to notification mails
- to go from invitation mail to calendar event
- to go to from task/lead notification to task/lead

do login first if user is not authenticated

**Comment by pschuele** [comment]
already implemented for events

**Comment by Gloirin** [comment]
**Comment posted by ketzaldev on 9 Dec 2010 15:01**

For sure, that feature is very important, because when you receive the notification mail, you need to search for the event.... it&#039;s not that friendly

**Comment by Gloirin** [comment]
**Comment posted by AirMike on 24 Oct 2012 08:18**

I really need this feature :(

Do You have any concept or idea how would You like to implement this so I could start with it :)

I did make this available for my custom module. It works but I&#039;m not satisfied with it and I would like it to implement generally.

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 24 Oct 2012 11:49**

do you use the tine email client (Felamimail) or an external client?

**Comment by Gloirin** [comment]
**Comment posted by AirMike on 24 Oct 2012 12:01**

I&#039;m not currently using Felamimail (do I need to use it for this?). 
I do use Tinebase Notifications (configuration is set in Setup (Smtp)) which are sending notifications and alarms, right?

**Comment by Gloirin** [comment]
**Comment posted by AirMike on 25 Oct 2012 10:10**

Any further info on this?

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 25 Oct 2012 10:24**

just for making sure, we are talking about the same thing:

you would like to add links to the notifications (like alarms, attender status changes, event invitations, ...) that, when clicked in an email client like outlook or thunderbird, open tine in the browser (show login first, if no session exists) and go to the calendar and open the event dialog for this event.

is that what you like to do?

we did some experiments in the past with a firefox plugin that the user has to install to make this work. but sadly the project has been sleeping since a year ... :(

**Comment by Gloirin** [comment]
**Comment posted by AirMike on 25 Oct 2012 10:42**

Exactly to the word :)

I would like to do this without any plugin.

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 25 Oct 2012 12:14**

ok, we see two problems when doing this without a plugin:

- if a tine window is already open, this can&#039;t be detected. this would open a new mainwindow every time
- popup-blocker: event dialog can&#039;t be opened because of this, you would need to disable this first
- initial registry loading and popup creation is slow

one idea about implementing this:

adding &quot;deep links&quot; to tine, with urls containing application, controller, action and id, like this:

https://yourtine.tld/calendar/(get?)/event/abcd1243554334

we could send you the incomplete (firefox) plugin if you like.

**Comment by Gloirin** [comment]
**Comment posted by AirMike on 25 Oct 2012 12:41**

I agree that without plugin there are problems You mentioned but requiring firefox as browser and installing additional plugin can be a bigger problem :) 

I think that detecting opened tine window is not required if opened window from link is not a &quot;full featured&quot; tine. By this I mean that we should initialize only things required for opening and editing requested record from &quot;deep link&quot; (no need for main screen, reducing registry fetch and similar stuff).

Is this possible?

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 25 Oct 2012 20:02**

i think that would be possible. but i&#039;m not sure what will happen, when the event is saved. perhaps it would only allow a read-only view of the event because the saving requires a main screen. but maybe we can find a solution for this.

@cornelius: what do you think?

**Comment by Gloirin** [comment]
**Comment posted by cweiss on 26 Oct 2012 09:12**

calendar can&#039;t save events w.o. mainscreen yet, and i think its hard to implement it.
Of course we can go this direction, but it would be quite a lot of work.

An other idea which came into my mind for mainwindow detection is to work with coockies or session storage polling. This way the windows can communicate even w.o. a parent/child relationship.

So the proceudure would be the following:
1. A &quot;deep link&quot; opens a new window wich loads index.php. 
2. The deeplink is parsed in the javascript bootstrap process.
3.1 If a mainwindow could be found, the new window becomes an edit dialog
3.2 otherwise it becomes a main window
3.2.1 the new main window queries for login
3.2.2 on sucessfull login the requested edit dialog is opened in a new window 
3.2.3 if a popup blocker got detected we open a messagebox with:
- please deactivate poupblocker AND click here to open dialog


The last remaining problem would be the resizeing problem for 3.1
If you open a link e.g. from outlook, you don&#039;t have an influence to the window size of the openend window. If this window than becomes a edit dialog, the dialogues just look ugly. window resizing per javascript is prohibited by most browsers.

And of course we have to implement cross window communication based on sessionStorage or coockies. But this would be usefull for tine anyway and be welcomed very much!

**Comment by Gloirin** [comment]
**Comment posted by cweiss on 26 Oct 2012 09:21**

for the session storage approach polling would&#039;t be necessary as there is a storage event

http://html5demos.com/storage-events

**Comment by Gloirin** [comment]
**Comment posted by AirMike on 10 Nov 2012 17:36**

Some help implementing this could really come in handy :)
I understand the procedure and I see that all work could be done in tineInit.js but I&#039;m not sure where and how locale storage comes into play.
I presume that when mainWindow is created it should store some data (only string are supported) in locale storage that opened window can read and act on it.
Also it would be nice if mainWindow is found that opened window doesn&#039;t have to load scripts of ALL modules but only scripts of Tinebase, Addressbook and calling module.
I really appreciate Your feedback and all discussions.

**Comment by Gloirin** [comment]
**Comment posted by AirMike on 14 Nov 2012 09:29**

Hmmm, how new window can become popup of main window and be able to close it with javascript?

**Comment by Gloirin** [comment]
**Comment posted by pschuele on 27 Nov 2012 11:37**

hi AirMike,

sorry for the delay, but cornelius is ill and will be back in the next week. he has already some ideas about the implementation and i hope he adds some kind of draft here.

**Comment by Gloirin** [comment]
**Comment posted by AirMike on 27 Nov 2012 14:31**

Cool, thank You.

**Comment by Gloirin** [comment]
**Comment posted by cweiss on 5 Dec 2012 11:16**

The point is: with session storage, all windows from one domain can have a common storage even if they don&#039;t have a parent/child relation.

Atm. the main window stores a &#039;isMainWindow&#039; property. In the bootstrap process we search down the parent tree for a mainWindow. If we find it we use its registry AND windowMgr to get our task, if not we become mainWindow ourself. More over in most cases we use the objects from the registry/mainWindow directly, this means javascript arrays and objects are shared between multiple javascript contexts. This causes a lot of problems, starting with IE, but also having cleanup issues in modern browsers.

So in future, our cross window communication should always be string based (json encoded). (This is also vaid for events etc. but at the moment we only talk about registry and sub window initialisation). This perfectly fits with the fact that local storage only stores strings (and by the way also with the fact, that HTML 5 workers also only communicate strings!)


So let&#039;s try to separate things out:

1. We need some kind of a string based, cross window accessible and non persistent/session only storage. (We might find a library for this.)

2. Based on this storage we need to have &quot;clients&quot; query data and to retrieve events on certain data changes to do our Messaging.

3. We need to adopt our &quot;window management&quot;. Windows need to get their tasks out of the storage from 1

4. We need to adopt our cross window registry / manager schema. It has to be based on 1 and 2.

5. We need to adopt our cross window event schema. It should be Message bus based. publisher send data to (1), Subscribers filter events of 2

6. With the help of 1-5 we can solve the initial problem: Windows opened e.g. from thunderbrid can get their needed information like registry etc. from 4 and talk to the main app through 5

Is the overall idea clear so far?

**Comment by Gloirin** [comment]
**Comment posted by AirMike on 6 Dec 2012 07:44**

Yes, I understand what we&#039;re trying to achieve and I think I saw some library for cross window communication (I&#039;ll try to look up again).
There is one problem, mainly for FF, and that is closing window because FF allows closing only windows that are opened via javascript and have opener.

P.S. maybe this library can help. http://easyxdm.net/wp/

**Comment by Gloirin** [comment]
**Comment posted by cweiss on 24 Jul 2014 09:26**

Some research of the current browser support:

http://caniuse.com/#search=sharedWorker
- this would be ideal, but
- bad browser support, why was it removed from safari?

http://caniuse.com/#search=sessionStorage
- no cross tab/window access/events
- not usable by any means

http://caniuse.com/#search=localStorage
- persistent storage, how to empty? 
- how to check that session is alive?
- Always use same key for messaging?

http://caniuse.com/#search=postMessage
- no cross window events
- bad IE support for IE &lt; 11

http://caniuse.com/#search=channel-messaging
- lacks the cross window events
- No FF support but IE support IE &gt;=10


To sum up:
* SharedWorker would be ideal, but their support is too bad atm.
* localStorage seems the way to got, but we need to find a way to validate/empty it
